### PR TITLE
Request Proxy in CF Worker

### LINKED ISSUE #473: Create a CF Worker to use it as a proxy
The goal here is to make http requests outside GraphQL Hive's network.

### PR DESCRIPTION
**Pull request by kamilkisiela**
The idea here is to run all outgoing communication (provided by users) within CF Worker to isolate it and disallow access to the cluster network.

Closes #473

### CODE REVIEW

File: packages/services/proxy-worker/src/dev.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  app.get(
+    '/_readiness',
+    () =>
+      new Response(null, {
+        status: 200,
+      })
+  );
+
+  app.get('*', (request: Request) => handleRequest(request, isSignatureValid));
```
Comment (ardatan) [inline packages/services/proxy-worker/src/dev.ts:22]:
```suggestion
  app.all('*', (request: Request) => handleRequest(request, isSignatureValid));
```

File: packages/services/proxy-worker/src/dev.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  app.post('*', (request: Request) => handleRequest(request, isSignatureValid));
```
Comment (ardatan) [inline packages/services/proxy-worker/src/dev.ts:23]:
```suggestion
```

File: packages/services/proxy-worker/src/handler.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * gatherResponse awaits and returns a response body as a string.
+ */
+async function gatherResponse(response: Response) {
+  const contentType = response.headers.get('content-type') || '';
+
+  if (contentType.includes('application/json')) {
+    return JSON.stringify(await response.json());
+  } else if (contentType.includes('application/text')) {
+    return response.text();
+  } else if (contentType.includes('text/html')) {
```
Comment (ardatan) [inline packages/services/proxy-worker/src/handler.ts:14]:
```suggestion
  } else if (contentType.startsWith('text/')) {
```

File: packages/services/proxy-worker/src/handler.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+          headers: parsedRequest.headers,
+        }
+      : {
+          method: 'POST',
+          body: parsedRequest.body,
+          headers: parsedRequest.headers,
+        };
+  const response = await fetch(parsedRequest.url, init);
+  const text = await gatherResponse(response);
+  return new Response(text, {
```
Comment (ardatan) [inline packages/services/proxy-worker/src/handler.ts:41]:
```suggestion
  return new Response(response.body, {
```
We can pass the body as-is.

**Comment by kamilkisiela** [comment]
I tested it on staging and I also prepared the Cloudflare DNS part.

File: packages/services/broker-worker/tests/borker.spec.ts [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        'X-Key': 'key',
+        accept: 'text/plain',
+      },
+    }),
+  });
+
+  const response = await handleRequest(request, SignatureValidators.Real);
+  expect(response.status).toBe(500);
+  expect(await response.text()).toBe('Internal Server Error');
+});
```
Comment (n1ru4l) [inline packages/services/broker-worker/tests/borker.spec.ts:339]: file name contains typo

Comment (kamilkisiela) [reply packages/services/broker-worker/tests/borker.spec.ts:339]: borker XD 🐶 bork bork

**Pull request merged**